src/ProductConstructionService/ProductConstructionService.ReproTool/ReproTool.cs

```diff
@@ -0,0 +1,321 @@
+
+        string vmrBranch, productRepoUri, productRepoBranch;
+        bool isForwardFlow;
+        // Check if it's forward flow
+        if (!string.IsNullOrEmpty(subscription.TargetDirectory))
+        {
+            vmrBranch = subscription.TargetBranch;
+            productRepoUri = subscription.SourceRepository;
+            productRepoBranch = defaultChannel.Branch;
+            isForwardFlow = true;
+        }
+        // Backward flow
+        else
+        {
+            vmrBranch = defaultChannel.Branch;
+            productRepoUri = subscription.TargetRepository;
+            productRepoBranch = subscription.TargetBranch;
+            isForwardFlow = false;
+        }
+        var productRepoForkUri = ProductRepoFormat + productRepoUri.Split('/').Last();
+        logger.LogInformation("Reproducing subscription from {sourceRepo} to {targetRepo}",
+            isForwardFlow ? productRepoForkUri : VmrForkUri,
+            isForwardFlow ? VmrForkUri : productRepoForkUri);
+
+        logger.LogInformation("Preparing test branch in repo {repo} from branch {branch}", VmrForkUri, vmrBranch);
+        await using var vmrTmpBranch = await PrepareVmrForkAsync(vmrBranch, productRepoUri, productRepoForkUri);
+
+        logger.LogInformation("Preparing product repo fork {productRepoFork}, branch {branch}", productRepoForkUri, productRepoBranch);
+        await using var productRepoTmpBranch = await PrepareProductRepoForkAsync(productRepoUri, productRepoForkUri, productRepoBranch);
+
+        // Find the latest commit in the source repo to create a build from
+        string sourceRepoSha;
+        (string sourceRepoName, string sourceRepoOwner) = GitRepoUrlParser.GetRepoNameAndOwner(subscription.SourceRepository);
+        if (build != null)
+        {
+            sourceRepoSha = build.Commit;
+        }
+        else if (string.IsNullOrEmpty(options.Commit))
+        {
+            var res = await ghClient.Git.Reference.Get(sourceRepoOwner, sourceRepoName, $"heads/{defaultChannel.Branch}");
+            sourceRepoSha = res.Object.Sha;
+        }
+        else
+        {
+            // Validate that the commit actually exists
+            try
+            {
+                await ghClient.Repository.Commit.Get(sourceRepoOwner, sourceRepoName, options.Commit);
+            }
+            catch (NotFoundException)
+            {
+                throw new ArgumentException($"Commit {options.Commit} doesn't exist in repo {subscription.SourceRepository}");
+            }
+            sourceRepoSha = options.Commit;
+        }
+
+        var channelName = Guid.NewGuid().ToString();
+        logger.LogInformation("Creating test channel {channelName}", channelName);
+        await using var channel = await darcProcessManager.CreateTestChannelAsync(channelName);
+
+        logger.LogInformation("Creating test build");
```

Comment:
Same, I'd put it inside `CreateBuildAsync`. We could also log the build number/id whatever to be able to query it later during testing.